node/src/tasks/bsp_move_bucket.rs

```diff
@@ -0,0 +1,184 @@
+use anyhow::anyhow;
+use sc_tracing::tracing::*;
+
+use shc_actors_framework::event_bus::EventHandler;
+use shc_blockchain_service::{
+    commands::BlockchainServiceInterface,
+    events::{MoveBucketAccepted, MoveBucketExpired, MoveBucketRejected, MoveBucketRequested},
+};
+use shc_file_transfer_service::commands::FileTransferServiceInterface;
+
+use crate::services::handler::StorageHubHandler;
+use crate::tasks::{BspForestStorageHandlerT, FileStorageT};
+
+const LOG_TARGET: &str = "bsp-move-bucket-task";
+
+const MOVE_BUCKET_ACCEPTED_GRACE_PERIOD_SECONDS: u64 = 4 * 60 * 60; // 4 hours
+
+/// Task that handles the [`MoveBucketRequested`], [`MoveBucketAccepted`], [`MoveBucketRejected`]
+/// and [`MoveBucketExpired`] events from the BSP point of view.
+pub struct BspMoveBucketTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: BspForestStorageHandlerT,
+{
+    storage_hub_handler: StorageHubHandler<FL, FSH>,
+}
+
+impl<FL, FSH> Clone for BspMoveBucketTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: BspForestStorageHandlerT,
+{
+    fn clone(&self) -> BspMoveBucketTask<FL, FSH> {
+        Self {
+            storage_hub_handler: self.storage_hub_handler.clone(),
+        }
+    }
+}
+
+impl<FL, FSH> BspMoveBucketTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: BspForestStorageHandlerT,
+{
+    pub fn new(storage_hub_handler: StorageHubHandler<FL, FSH>) -> Self {
+        Self {
+            storage_hub_handler,
+        }
+    }
+}
+
+/// Handles the [`MoveBucketRequested`] event.
+///
+/// This event is triggered when an user requests to move a bucket to a new MSP.
+/// As a BSP, we need to allow the new MSP to download the files we have from the bucket.
+impl<FL, FSH> EventHandler<MoveBucketRequested> for BspMoveBucketTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: BspForestStorageHandlerT,
+{
+    async fn handle_event(&mut self, event: MoveBucketRequested) -> anyhow::Result<()> {
```

Comment:
Shouldn't this check if this BSP even has files from that bucket?

Reply:
I mean it's not mandatory, but it would be a nice to have so that we don't clutter the allow list. Anyway, there shouldn't be that many bucket changes ideally.

Reply:
In order to do that we need to have an index from `BucketId` to `FileKey`s. This is 1 of the reasons why I opted to have allowlists for buckets and let the BSP check the `FileKey,BucketId` requested pair. To fix just this we can also add an index just for all `BucketId`s stored.